docs/parameterized-function-expressions.md

```diff
@@ -0,0 +1,179 @@
+
+## Drawbacks
+
+The syntax proposed in this RFC would close the door on the ability to add a bitwise left-shift operator to Luau. This is because this would make the following usage ambiguous:
+
+```lua
+-- Assuming we have right-shift, because there's no way we add one without the other
+return f<<a, b>>(c)
+```
+
+This is either a parameterized function call of `f(c)` with `a` and `b` as types, or a return of two values--`f << a`, and `b >> (c)`.
+
+Because there would be no ability to add bitwise shifts to the language, it is unlikely that any bitwise operators would be added, as it would be a stark omission.
+
+With that said, the ability to perform these kinds of shifts is already available through the `bit32` library, and [Luau's compatibility page](https://luau.org/compatibility#lua-53) has the following words on the topic in reference to their availability in Lua 5.3:
+
+> If integers are taken out of the equation, bitwise operators make less sense, as integers aren’t a first class feature; additionally, bit32 library is more fully featured (includes commonly used operations such as rotates and arithmetic shift; bit extraction/replacement is also more readable). Adding operators along with metamethods for all of them increases complexity, which means this feature isn’t worth it on the balance.
+
+The syntax is also unfamiliar not just in the context of Luau but in the context of programming languages in general--`<<` and `>>` as type parameter specification is completely unique to this call alone.
+
+## Alternatives
+
+### Syntax
+
+First, what we absolutely cannot pick:
+
+- `f<T>()` is the choice of C++, but is ambiguous with `f<a, b>(c)` in the same way we mentioned with bitwise operators above. C++ only disambiguates this with an extra pass using type information that would dramatically inflate the complexity of all Luau parsers.
+- `f::<T>()`, named the "turbofish", is the choice of Rust. This is not strictly ambiguous, but it would require infinite lookahead as it conflicts with casting generic functions. In `local x = f::<T>(() -> T) -> ()`, which is valid code today, we cannot know this is a type assignment until we get to the first `)`, as until that point we might have code that looks like `local x = f::<T>((x))`. This is not a cost we would like to add to the parser.
+
+For unambiguous alternatives, `f.<T>` and `f:<T>` are among the most discussed. This would look like:
+
+```lua
+local moneyBinding = React.createBinding.<number>()
+
+-- or...
+local moneyBinding = React.createBinding:<number>()
+```
+
+The downside of these is that they blur the lines between runtime and static, in the sense that `React.createBinding.` starts out as a runtime concept, followed by the purely static `<number>`. As for `:`, it carries the baggage of `x:y()` which will perform a runtime mutation of the function call in the form of adding on `self`.

```

Comment:
I would not describe this as a runtime mutation. It's very confusing. There's no mutation happening, `x:y()` is just syntactic sugar for `x.y(x)`.

Reply:
Yeah I'm not sure the right way to say what I want to say--what I want to say is that `:` is inherently more "interesting" than `.`